src/browser.cc

```diff
@@ -0,0 +1,369 @@
+IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+SOFTWARE.
+*******************************************************************************/
+
+#include "helpers.h"
+
+#include "insights.h"
+CEREAL_REGISTER_TYPE(insight::MutualInformation);
+
+#include "html.h"
+
+#include "../../Current/Bricks/dflags/dflags.h"
+#include "../../Current/Bricks/file/file.h"
+#include "../../Current/Bricks/net/api/api.h"
+#include "../../Current/Bricks/strings/util.h"
+#include "../../Current/Bricks/waitable_atomic/waitable_atomic.h"
+
+DEFINE_int32(port, 3000, "Port to spawn the browser on.");
+DEFINE_string(route, "/", "The route to serve the browser on.");
+DEFINE_string(output_uri_prefix, "http://localhost:3000", "The prefix for the URI-s output by the server.");
+
+DEFINE_string(input, "data/insights.json", "Path to the file containing the insights to browse.");
+DEFINE_string(id_key, "you_are_awesome", "The URL parameter name containing smart session token ID.");
+
+using bricks::strings::Printf;
+using bricks::strings::FromString;
+using bricks::strings::ToString;
+using bricks::FileSystem;
+using bricks::WaitableAtomic;
+
+std::string RandomString(const size_t length = 8) {
+  std::string s;
+  for (size_t i = 0; i < length; ++i) {
+    s += 'a' + rand() % 26;
+  }
+  return s;
+}
+
+struct TopLevelResponse {
+  std::string SMART_HTML_BROWSE_URI;
+  std::string HTML_BROWSE_URI;
+  std::string SMART_BROWSE_URI;
+  size_t total;
+  std::string browse_uri;
+  std::string browse_all_uri;
+  TopLevelResponse(size_t i) : total(i) {
+    if (total) {
+      const std::string P = FLAGS_output_uri_prefix + FLAGS_route;
+      SMART_BROWSE_URI = P + "smart";
+      HTML_BROWSE_URI = P + "?id=1&html=yes";
+      SMART_HTML_BROWSE_URI = P + "smart?html=yes";
+      browse_uri = P + "?id=1";
+      browse_all_uri = P + "?id=all";
+    }
+  }
+  template <typename A>
+  void serialize(A& ar) {
+    ar(CEREAL_NVP(SMART_HTML_BROWSE_URI),
+       CEREAL_NVP(HTML_BROWSE_URI),
+       CEREAL_NVP(SMART_BROWSE_URI),
+       CEREAL_NVP(total),
+       CEREAL_NVP(browse_uri),
+       CEREAL_NVP(browse_all_uri));
+  }
+};
+
+struct InsightResponse {
+  std::string current_uri;  // Permalink.
+  std::string prev_uri;
+  std::string next_uri;
+  std::set<std::string> tags;
+  double score;
+  std::string description;
+  std::unique_ptr<insight::AbstractBase> insight;
+  InsightResponse() = default;
+  InsightResponse(const InsightsOutput& input, size_t index) { Prepare(input, index); }
+  void Prepare(const InsightsOutput& input, size_t index) {
+    const std::string P = FLAGS_output_uri_prefix + FLAGS_route;
```

Comment:
It's not `P`, it's `url` (more precisely the `output_uri_prefix` is actually a `base_url`, not a `uri`).